### PR TITLE
[Fix] Fix DVB bug (Multiple-line subtitle; Missing last line)

### LINKED ISSUE #840: [BUG] unable to extract multiple lines from DVB-sub using OCR
CCExtractor version (using the --version parameter preferably) : **0.85** _and_ **Git commit 18584259447a145a1b8c1cae6733223393dfb4f1**

**In raising this issue, I confirm the following (please check boxes, eg [X]):**

- [x] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that the bug-fix I am reporting can be replicated, or that the feature I am suggesting isn't already present.
- [x] I have checked that the issue I'm posting isn't already reported.
- [x] I have checked that the issue I'm porting isn't already solved and no duplicates exist in [closed issues](https://github.com/CCExtractor/ccextractor/issues?q=is%3Aissue+is%3Aclosed) and in [opened issues](https://github.com/CCExtractor/ccextractor/issues)
- [x] I have checked the pull requests tab for existing solutions/implementations to my issue/suggestion.
- [x] I have used the latest available version of CCExtractor to verify this issue exists.

I found https://github.com/CCExtractor/ccextractor/issues/392 which seems like the same issue, but it was closed in mid-2016, and I was able to reproduce the issue on a much newer version than that (including a Git build from today.) Therefore I think it's OK to file this.

Fortunately, the latest Git version seems to have addressed another bug I was going to file. Although the "something messy" instrumentation in you added would have to be filtered out, disabled, or sent to stderr before I could use this version.

**My familiarity with the project is as follows (check one, eg [X]):**

- [ ] I have never used CCExtractor.
- [ ] I have used CCExtractor just a couple of times.
- [x] I absolutely love CCExtractor, but have not contributed previously.
- [ ] I am an active contributor to CCExtractor.

**Necessary information**
- Is this a regression (did it work before)? [x] NO | [ ] YES - *please specify the last known working version*
- What platform did you use? [ ] Windows - [x] Linux - [ ] Mac
64-bit CentOS 7.3
- What where the used arguments? `ccextractor -stdout -quiet -nofc -nodvbcolor  ccextractor_bugs_allcaps_29fps_leftjustify.m2ts`

**Video links**

https://s3-us-west-2.amazonaws.com/ccextractor-dvbsub-bugreports/ccextractor_bugs_allcaps_29fps_leftjustify.m2ts

I think I have set the permissions correctly on this S3 bucket, let me know if you can't download it.

The DVB-sub captions in this video display perfectly in VLC (tested verison 2.2.4.) There is a burned in timecode on the video to help you judge the timing, and it seems dead-on to me in VLC.

**Additional information**

There are three problems with ccextractor's output from this file:

- Timing is slightly off (really not a huge deal, it's only off by 10-30 ms)
- The last caption (the twenty-fifth caption) is missing from the output completely
- Only the first line of text from each caption is shown (this is the biggest problem)

When I use spupng output and run the tesseract command line program (version 3.04 tested) on the PNG images, the text is detected just fine, so I don't think it's a limitation of tesseract itself.

If you're curious about the bug that was fixed with the latest Git version, it's that several captions were missing altogether when extracted from this file:  
https://s3-us-west-2.amazonaws.com/ccextractor-dvbsub-bugreports/big_buck_bunny_eac3_4.m2ts  
even though they played just fine in VLC. This bug is fixed now, but you could use the video in your regression tests if you want.

### PR DESCRIPTION
**Pull request by harrynull**
Please prefix your pull request with one of the following: **[FEATURE]** **[FIX]** **[IMPROVEMENT]**.

**In raising this pull request, I confirm the following (please check boxes):**

- [X] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [X] I have checked that another pull request for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**My familiarity with the project is as follows (check one):**

- [ ] I have never used CCExtractor.
- [ ] I have used CCExtractor just a couple of times.
- [ ] I absolutely love CCExtractor, but have not contributed previously.
- [X] I am an active contributor to CCExtractor.

---
Fix #840 
- Fix the bug that only the first line of text from each caption is shown.
- Fix the bug that the last caption is missing.
- The shifting time isn't really a problem because it is indeed shifting a little bit in the video and the conversion from fps->1000 will also cause some precision problems.

### CODE REVIEW

File: src/lib_ccx/ccx_encoders_srt.c [hunk 1]
```diff
@@ -83,6 +83,7 @@ int write_stringz_as_srt(char *string, struct encoder_ctx *context, LLONG ms_sta
 
 int write_cc_bitmap_as_srt(struct cc_subtitle *sub, struct encoder_ctx *context)
 {
+    mprint("\nNow subtitle %x is really written into file\n", sub);
```
Comment (cfsmp3) [inline src/lib_ccx/ccx_encoders_srt.c]: Debug messages should use the debug function, not mprint()...

File: src/lib_ccx/dvb_subtitle_decoder.c [hunk 2]
```diff
@@ -1551,6 +1551,9 @@ static int write_dvb_sub(struct lib_cc_decode *dec_ctx, struct cc_subtitle *sub)
 				rect->ocr_text = ocr_str;
 			else
 				rect->ocr_text = NULL;
+            if (ccx_options.dvb_debug_traces_to_stdout) {
+                mprint("\nOCR Result: %s\n", rect->ocr_text ? rect->ocr_text : "NULL");
```
Comment (cfsmp3) [inline src/lib_ccx/dvb_subtitle_decoder.c]: Same thing... plus debug options are controlled with a mask, not with independent variables.

File: src/lib_ccx/general_loop.c [hunk 3]
```diff
@@ -887,6 +887,7 @@ int general_loop(struct lib_ccx_ctx *ctx)
     }
 
 	end_of_file = 0;
+    struct encoder_ctx *enc_ctx = NULL;
```
Comment (cfsmp3) [inline src/lib_ccx/general_loop.c]: Incorrect identation (unless github is not displaying it right)

File: src/lib_ccx/ccx_encoders_srt.c (hunk 1 above)
Comment (harrynull) [reply src/lib_ccx/ccx_encoders_srt.c]: Sorry, it's supposed to be temporary but I forgot to remove them.

File: src/lib_ccx/dvb_subtitle_decoder.c (hunk 2 above)
Comment (harrynull) [reply src/lib_ccx/dvb_subtitle_decoder.c]: I did not add the option. It is controlled by parameter `-debugdvbsub` and is used many places in the code already.

File: src/lib_ccx/general_loop.c [hunk 4]
```diff
@@ -1135,6 +1135,14 @@ int general_loop(struct lib_ccx_ctx *ctx)
 		if (ccx_options.send_to_srv)
 			net_check_conn();
 	}
+
+    // Process the last subtitle
```
Comment (anshul1912) [inline src/lib_ccx/general_loop.c]: not a good way to fix it, should remove encode_sub from dvb_sub_decoder

File: src/lib_ccx/dvb_subtitle_decoder.c [hunk 5]
```diff
@@ -419,7 +419,7 @@ static void delete_regions(DVBSubContext *ctx)
  *                       it could be -1 if not found in PMT.
  * @param ancillary_id ancillary-page_id found in Subtitle descriptors
  *                     associated with     subtitle stream in the    PMT.
- *                       it could be -1 if not found in PMT.
+ *                       it could be -1 if not found in PMT.
```
Comment (anshul1912) [inline src/lib_ccx/dvb_subtitle_decoder.c]: why added extra space above

**Pull request merged**